kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/sink/seek/IndexManager.scala

```diff
@@ -0,0 +1,184 @@
+    ))
+      .leftMap(e => new NonFatalS3SinkError("Couldn't retrieve listing", e.exception))
+      .flatMap(
+        indexes =>
+          if (indexes.size > maxIndexes) {
+            logAndReturnMaxExceededError(topicPartition, indexes)
+          } else if (indexes.isEmpty && fallbackSeeker.nonEmpty) {
+            fallbackOffsetSeek(fallbackSeeker.get, topicPartition, fileNamingStrategy, bucketAndPrefix)
+          } else {
+            {
+              for {
+                validIndex <- scanIndexes(bucketAndPrefix, indexes)
+                indexesToDelete = indexes.filterNot(validIndex.contains)
+                _ <- storageInterface.deleteFiles(bucketAndPrefix.bucket, indexesToDelete)
+                offset <- indexToOffset(validIndex)
+              } yield {
+                logger.info("[{}] Seeked offset {} for TP {}", sinkName, offset, topicPartition)
+                offset
+              }
+            }.leftMap {
+              case err: FileLoadError =>
+                val logLine = s"File load error while seeking: ${err.message()}"
+                logger.error(s"[{}] {}",sinkName, logLine, err.exception)
+                new NonFatalS3SinkError(logLine, err.exception)
+              case err: FileDeleteError =>
+                val logLine = s"File delete error while seeking: ${err.message()}"
+                logger.error(s"[{}] {}",sinkName, logLine, err.exception)
+                new NonFatalS3SinkError(logLine, err.exception)
+              case err: Throwable =>
+                val logLine = s"Error while seeking: ${err.getMessage}"
+                logger.error(s"[{}] {}",sinkName, logLine, err)
+                new NonFatalS3SinkError(logLine, err)
+            }
+          }
+      )
+  }
+
+  /**
+   * Given a bucket and a list of files, attempts to load them to establish the most recent valid index
+   *
+   * @param bucketAndPrefix Remote S3 Root Location of the sink files
+   * @param indexFiles      List of index files
+   * @return either a FileLoadError or an optional string containing the valid index file of the offset
+   */
+  def scanIndexes(bucketAndPrefix: RemoteS3RootLocation, indexFiles: List[String]): Either[FileLoadError, Option[String]] = {
+    var latest: Either[FileLoadError, Option[String]] = None.asRight
+
+    indexFiles
+      .reverse
+      .takeWhile {
+        idxFileName: String =>
+          latest = {
+            for {
+              targetFileName <- storageInterface.getBlobAsString(bucketAndPrefix.withPath(idxFileName))
+              targetBucketAndPath = bucketAndPrefix.withPath(targetFileName)
+              pathExists <- storageInterface.pathExists(targetBucketAndPath)
+            } yield {
+              if (pathExists) Some(idxFileName) else Option.empty[String]
+            }
+          }
+        latest.isRight && latest.right.exists(_.isEmpty)
+      }
+
+      latest
```

Comment:
var field :)
Let's make it a bit more functional. 
I suggest we list all the indexes and split the return into:
1. files with a target file which exists
2. files with a target file which does not exist (this would need to be delete somewhere else)

the point1 list can then be iterated over by the caller and delete anything but last. and this we can do by ordering in descending order and delete the `.tail` of the list

Reply:
It's best to break in smaller functions as well.
The function which takes a the files and reads the content should be one function. 

The result of this function (whatever the shape, it needs to have a index path and Option[target path]) will be the input to another function which can identify the files which need to be deleted and also another function which can identify the latest (let's not optimise for performance - iterating over a few items list it will be fast and there's a lot of I/O in the connector to worry about performance rather than CPU cycles and a bit of extra memory allocation)

Reply:
> var field :) Let's make it a bit more functional. I suggest we list all the indexes and split the return into:
> 
> 1. files with a target file which exists
> 2. files with a target file which does not exist (this would need to be delete somewhere else)
> 
> the point1 list can then be iterated over by the caller and delete anything but last. and this we can do by ordering in descending order and delete the `.tail` of the list

1. I feel bad/guilty for using the var.  I know there must be some functional alternative
2. I considered similar to this algorithm (above) but the problem with that, given 100 index files (say 0…100) then you are:
    - loading 100 index files to retrieve the source
    - attempting to load 100 target files
Whereas it is possible you only need to load a single file (/100), if this is valid then you can just move on and delete all the rest without 99 * IO to S3 which is over a network and relatively slow.

Reply:
Please clarify my understanding:

1. we have the file to upload and we know the last offset
2. we write the index file (the name is the offset value - we do it per partition ). the file contains the target file path we are about to upload.
3. we upload the file. if successful then the index file has a valid reference. otherwise the index file is "orphaned".

When we extract the index
* during topic-partition re-assignment
* during a write (we want to avoid writing orphaned files)

Therefore listing all indexes for the topic-partition should have at most 2 files, in most scenarios(success) it's one. We have two files if we  reached step 2 but 3 failed.

So we list the files and read their content to see where they point two and if the referenced file exists. if the reference file does not exist, the index file:
1. is not included it when calculating the latest
2. needs to be deleted


So if the above understanding is correct, then we should not rely on only picking the latest index file and iterate over all available. Like i said we should not have more than 2 if we do the cleanup. Thoughts?

Reply:
This is absolutely correct.

However... if there is a problem with file deletion, we do not stop the sink of throw an error, and assume this will be cleaned up later.

An example of such a problem is that the S3 permissions are set to allow the creation of files but not deletion.  Therefore delete operations will fail but we continue on regardless.

Whilst it is an unusual case, is it not worth ensuring the code is still efficient?